src/parser/bom.rs

```diff
@@ -0,0 +1,148 @@
+//! A parser for encoding detection using BOM and heuristics.
+
+/// A result of feeding data into [`BomParser`].
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
```

Comment:
FYI, one of the sticking points on the encoding PR is/was that if you wrap the input stream of bytes in an decoding wrapper, then the BOM will always be translated to UTF-8 before the XML parser gets a chance to take a look at it.

It's a thoroughly fixable issue, but it means you have to either poke a hole in the decoder to get a look at raw bytes and set the encoding manually from the outside, or move the encoding detection heuristics to the inner decoding wrapper, or read a couple of bytes, determine the encoding, set the encoding on the decoder from outside, and then replay the bytes previously read.